src/Microsoft.AspNetCore.Mvc.Core/Formatters/TextInputFormatter.cs

```diff
@@ -84,21 +134,27 @@ protected Encoding SelectCharacterEncoding(InputFormatterContext context)
 
             if (SupportedEncodings?.Count == 0)
             {
-                var message = Resources.FormatTextInputFormatter_SupportedEncodingsMustNotBeEmpty(nameof(SupportedEncodings));
+                var message = Resources.FormatTextInputFormatter_SupportedEncodingsMustNotBeEmpty(
+                    nameof(SupportedEncodings));
                 throw new InvalidOperationException(message);
             }
 
-            var request = context.HttpContext.Request;
-
-            var requestEncoding = request.ContentType == null ? null : MediaType.GetEncoding(request.ContentType);
-            if (requestEncoding != null)
+            var requestContentType = context.HttpContext.Request.ContentType;
+            var requestEncoding = requestContentType == null ?
+                default(StringSegment) :
+                new MediaType(requestContentType).Charset;
+            if (requestEncoding.HasValue)
             {
+                string encodingName;
+                if (!EncodingAliases.TryGetValue(requestEncoding.Value, out encodingName))
+                {
+                    // Given name was not an encoding alias. Use the original value.
+                    encodingName = requestEncoding.Value;
+                }
+
                 for (int i = 0; i < SupportedEncodings.Count; i++)
                 {
-                    if (string.Equals(
-                        requestEncoding.WebName,
-                        SupportedEncodings[i].WebName,
-                        StringComparison.OrdinalIgnoreCase))
+                    if (string.Equals(encodingName, SupportedEncodings[i].WebName, StringComparison.OrdinalIgnoreCase))
```

Comment:
`encodingName` is a `string`. The code gets the `Value` (creates a substring) anyhow for the `TryGetValue()` a few lines earlier. But I can use another variable to create the substring once instead of twice in the `EncodingAliases` miss case. Missed fact `StringSegment` doesn't store its `Value` once read.